### PR TITLE
Drag bounds

### PR DESCRIPTION
**Pull request by aakamenov**
Hi emilk,

This PR is for #178 . Let me know if you see any issues.

Thanks

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 * `Context::set_pixels_per_point` to control the scale of the UI.
 * Add `Response::changed()` to query if e.g. a slider was dragged, text was entered or a checkbox was clicked.
 * Add support for all integers in `DragValue` and `Slider` (except 128-bit).
+* Add the option to restrict the dragging bounds of `Window` and `Area` to a specified area using `drag_bounds(rect)`.
```
Comment (emilk) [inline CHANGELOG.md:25]: Please move to under the `## Unreleased` section instead

File: egui/src/containers/area.rs [hunk 2]
```diff
@@ -130,13 +132,20 @@ impl Area {
         self.new_pos = Some(current_pos);
         self
     }
+
+    ///Constrain the area up to which the window can be dragged.
```
Comment (emilk) [inline egui/src/containers/area.rs:136]:
```suggestion
    /// Constrain the area up to which the window can be dragged.
```

File: egui/src/containers/window.rs [hunk 3]
```diff
@@ -204,6 +204,12 @@ impl<'open> Window<'open> {
         }
         self
     }
+
+    ///Constrain the area up to which the window can be dragged.
```
Comment (emilk) [inline egui/src/containers/window.rs:208]:
```suggestion
    /// Constrain the area up to which the window can be dragged.
```

File: egui/src/context.rs [hunk 4]
```diff
@@ -464,33 +464,41 @@ impl Context {
...
+    }
+
+    /// Constrain the position of a window/area
+    /// so it fits within the provided boundary.
+    pub(crate) fn constrain_window_rect_to_area(&self, window: Rect, area: Rect) -> Rect {
+        self.constrain_rect_impl(window, area)
+    }
 
-        if window.width() > screen.width() {
+    fn constrain_rect_impl(&self, window: Rect, mut area: Rect) -> Rect {
```
Comment (emilk) [inline egui/src/context.rs:479]: no need for both `constrain_rect_impl` and `constrain_window_rect_to_area`. Just keep `constrain_window_rect_to_area` and call it from `constrain_window_rect`

**Review by emilk — CHANGES_REQUESTED** [review]
Look good - just some small changes required!

File: CHANGELOG.md [hunk 5]
```diff
@@ -16,12 +16,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
   * Use arrow keys to adjust sliders and `DragValue`s.
 * egui will now output events when widgets gain keyboard focus.
   * This can be hooked up to a screen reader to aid the visually impaired
+* Add the option to restrict the dragging bounds of `Window` and `Area` to a specified area using `drag_bounds(rect)`.
 
 ### Fixed 🐛
 
 * Fixed secondary-click to open a menu
 
-
```
Comment (emilk) [inline CHANGELOG.md:24]: nit: revert

File: egui/src/containers/area.rs [hunk 6]
```diff
@@ -215,13 +226,23 @@ impl Prepared {
         &mut self.state
     }
 
+    pub(crate) fn drag_bounds(&self) -> Option<Rect> {
+        if self.drag_bounds.is_some() {
+            Some(self.drag_bounds.unwrap())
+        } else {
+            None
+        }
```
Comment (emilk) [inline egui/src/containers/area.rs:234]:
```suggestion
        self.drag_bounds
```

**Comment by aakamenov** [comment]
All done!

**Review by emilk — CHANGES_REQUESTED** [review]
almost there :)

File: egui/src/containers/area.rs (hunk 6 above)
Comment (aakamenov) [reply egui/src/containers/area.rs]: I really don't know what I was thinking! 🤕

**Comment by emilk** [comment]
Thank you @aakamenov !

**Pull request merged**